<h1 dir=rtl align=center style="line-height:200%;font-family:vazirmatn;color:#0099cc">
<font face="vazirmatn" color="#0099cc">
سیمان بازی
</font>
</h1>

<h2 dir=rtl align=right style="line-height:200%;font-family:vazirmatn;color:#0099cc">
<font face="vazirmatn" color="#0099cc">
مقدمه و صورت مسئله
</font>
</h2>


<p dir=rtl style="direction: rtl; text-align: justify; line-height:200%; font-family:vazirmatn; font-size:medium">
<font face="vazirmatn">
این مجموعه داده شامل فرمولاسیون‌های مختلف سیمان و مقاومت فشاری متناظر با آن‌هاست. هدف این بخش، پیش‌بینی مقاومت بتن بر اساس مواد اولیه و زمان عمل‌آوری آن است. با تحلیل این ویژگی‌ها، شما یک مدل برای پیش‌بینی دقیق مقاومت نهایی سیمان توسعه خواهید داد. هدف شما پیاده نمودن یک مدل رگرسیونی است.
</font>
</p>

<h2 dir=rtl align=right style="line-height:200%;font-family:vazirmatn;color:#0099cc">
<font face="vazirmatn" color="#0099cc">
معرفی مجموعه داده
</font>
</h2>

<p dir=rtl style="direction: rtl; text-align: justify; line-height:200%; font-family:vazirmatn; font-size:medium">
<font face="vazirmatn" size=3>
در فایل اولیه‌ی این سوال، یک پوشه با نام <code>Data</code> وجود دارد که شامل ۲ فایل <code>train.csv</code> و <code>test.csv</code>.
</font>
</p>

<p dir=rtl style="direction: rtl; text-align: justify; line-height:200%; font-family:vazirmatn; font-size:medium">
<font face="vazirmatn" size=3>

</font>
</p>


<p dir=rtl style="direction: rtl; text-align: justify; line-height:200%; font-family:vazirmatn; font-size:medium">
<font face="vazirmatn" size=3>
هرستون این مجموعه‌داده با توجه به ماهیت خود می‌تواند مقادیر عددی یا رشته‌ای داشته باشد. برای آشنایی بیشتر با مجموعه‌داده، پیشنهاد می‌شود هرستون به صورت جداگانه بررسی شود.</font>
</p>

<h4 dir=rtl align=right style="line-height:200%;font-family:vazir;color:#0099cc">
<font face="vazir" color="#0099cc">
معرفی مجموعه داده آموزشی (train.csv)
</font>
</h4>

<div dir="rtl">
<p dir=rtl style="direction: rtl; text-align: justify; line-height:200%; font-family:vazir; font-size:medium">
<font face="vazir" size=3>

فایل <code>train.csv</code> دارای `824` سطر و `12` ستون می‌باشد که ستون‌های آن به شرح زیر می‌باشد:

<center>
<div dir=rtl style="direction: rtl;line-height:200%;font-family:vazir;font-size:medium">
<font face="vazir" size=3>
    

| نام ستون | توضیح |
|:---: |:---: |
| `Cement ` | مقدار سیمان بر حسب کیلوگرم بر متر مکعب. سیمان ماده اصلی چسبنده در بتن است و مقدار آن مستقیماً با مقاومت نهایی مرتبط است.|
| `Blast Furn Slag` | مقدار سرباره کوره بلند بر حسب کیلوگرم بر متر مکعب. این ماده یک افزودنی سیمانی است که مقاومت بتن، به ویژه در بلندمدت، را افزایش می‌دهد.|
| `Fly Ash` |مقدار خاکستر بادی بر حسب کیلوگرم بر متر مکعب. این ماده افزودنی، که محصول جانبی احتراق زغال سنگ است، به بهبود کارایی و مقاومت بتن کمک می‌کند.|
| `Water ` | مقدار آب بر حسب کیلوگرم بر متر مکعب.|
| `Superplas` | مقدار فوق روان کننده بر حسب کیلوگرم بر متر مکعب.|
| `Coarse Ag` |مقدار سنگدانه درشت (شن یا سنگ خرد شده) بر حسب کیلوگرم بر متر مکعب.ی |
| `Fine Aggre` | مقدار سنگدانه ریز (ماسه) بر حسب کیلوگرم بر متر مکعب. این ماده فضاهای خالی بین سنگدانه‌های درشت را پر کرده و به صافی و تراکم بتن می‌افزاید.|
| `Age` | سن بتن در زمان آزمایش بر حسب روز.|
| `cement per water` |این ستون نسبت سیمان به آب را نشان می‌دهد.|
| `Cement Impurity Factor` |معیاری برای ناخالصی‌های موجود در سیمان را نشان می‌دهد.|
| `Cement Moisture Factor` |میزان رطوبت موجود در سیمان باشد. رطوبت سیمان می‌تواند بر نسبت مؤثر آب به سیمان و در نتیجه بر مقاومت بتن تأثیر بگذارد.|
| `Strength` |این متغیر هدف است که باید آن را پیش‌بینی کنید. این ستون مقاومت فشاری بتن را بر حسب مگاپاسکال نمایش می‌دهد.|


</font>
</div>
</center>
</div>

<h4 dir=rtl align=right style="line-height:200%;font-family:vazirmatn;color:#0099cc">
<font face="vazirmatn" color="#0099cc">
معرفی مجموعه داده آزمون (test.csv)
</font>
</h4>

<div dir="rtl">
<p dir=rtl style="direction: rtl; text-align: justify; line-height:200%; font-family:vazirmatn; font-size:medium">
<font face="vazirmatn" size=3>
فایل <code>test.csv</code> ستون‌هایی مشابه با فایل <code>train.csv</code> می‌باشد. همچنین در این فایل <code>206</code> سطر وجود دارد.
با این تفاوت که این فایل شامل ستون <code>Strength</code> نمی‌باشد.
</font>

<p dir=rtl style="direction: rtl; text-align: justify; line-height:200%; font-family:vazirmatn; font-size:medium">
<font face="vazirmatn" size=3>
هدف نهایی شما این است که با استفاده از یک مدل با استفاده از یادگیری ماشین یا یادگیری عمیق آموزش دهید تا بتواند مقدار مقاومت بتن را به ازای هر نمونه آزمایش پیش‌بینی کند.
</font>
</div>

<p dir=rtl style="direction: rtl; text-align: justify; line-height:200%; font-family:vazirmatn; font-size:medium">
<font face="vazirmatn" size=3>
در این قسمت کتاب‌خانه‌ها و ابزار مورد نیاز خود را <code>import</code> کنید و فایل داده‌ها را که در پوشه‌ی <code>Data</code> و که با نام‌های <code>train.csv</code> و <code>test.csv</code> ذخیره‌شده‌اند را بخوانید و وارد محیط کار خود کنید.
</font>
</p>

In [149]:
import pandas as pd
import numpy as np
train=pd.read_csv("Data/train.csv")
test=pd.read_csv("Data/test.csv")

In [150]:
train=train.drop(columns=['Cement Moisture Factor'])
test=test.drop(columns=['Cement Moisture Factor'])

<h2 dir=rtl align=right style="line-height:200%;font-family:vazirmatn;color:#0099cc">
<font face="vazirmatn" color="#0099cc">
پیش‌پردازش و مهندسی ویژگی
</font>
</h2>

<p dir=rtl style="direction: rtl; text-align: justify; line-height:200%; font-family:vazirmatn; font-size:medium">
<font face="vazirmatn" size=3>
        در این سوال شما می‌توانید از هر تکنیک پیش‌پردازش/مهندسی ویژگی دلخواهتان، استفاده کنید.
    <br>
    تکنیک‌هایی که استفاده می‌کنید به شکل مستقیم مورد ارزیابی توسط سامانه داوری قرار <b>نمی‌گیرند.</b> بلکه همه آن‌ها در دقت مدل شما تاثیر خواهند گذاشت؛ بنابراین هر چه پیش‌پردازش/مهندسی ویژگی بهتری انجام دهید تا دقت مدل بهبود پیدا کند، امتیاز بیشتری از این سوال کسب خواهید کرد.
    در این قسمت شما می‌توانید بخشی از داده‌ی موجود را برای اعتبارسنجی در نظر بگیرید.
</font>
</p>

In [151]:
# train['strength_class'] = pd.qcut(train['Strength'], q=10, labels=False)

In [152]:
# tempx=train.drop(columns=['strength_class',"Strength"])
# tempy=train['strength_class']

In [153]:
# def class_mean_imputer_fit(X, Y):

#     class_means = X.groupby(Y).mean()
#     return class_means

# def class_mean_imputer_transform(X, Y, class_means):

#     X_copy = X.copy()

#     for i, y_val in enumerate(Y):
#         if y_val in class_means.index:
#             X_copy.loc[i] = X_copy.loc[i].fillna(class_means.loc[y_val])
    
#     X_copy = X_copy.fillna(X.mean())

#     return X_copy, Y


# tempx = train.drop(columns=['strength_class', 'Strength'])
# tempy = train['strength_class']

# classmeans = class_mean_imputer_fit(tempx, tempy)
# x, y = class_mean_imputer_transform(tempx, tempy, classmeans)


In [154]:
# train = pd.concat([x, train['Strength']], axis=1)

In [155]:
train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 824 entries, 0 to 823
Data columns (total 11 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Cement                  824 non-null    float64
 1   Blast Furnace Slag      824 non-null    float64
 2   Fly Ash                 824 non-null    float64
 3   Water                   824 non-null    float64
 4   Superplasticizer        824 non-null    float64
 5   Coarse Aggregate        824 non-null    float64
 6   Fine Aggregate          824 non-null    float64
 7   Age                     824 non-null    int64  
 8   cement per water        824 non-null    float64
 9   Cement Impurity Factor  824 non-null    float64
 10  Strength                824 non-null    float64
dtypes: float64(10), int64(1)
memory usage: 70.9 KB


In [156]:
# from sklearn.preprocessing import MinMaxScaler

# def MinMaxScalarNorm(X,Y):
#   scaler = MinMaxScaler()
#   scaler.fit(X)
#   return pd.DataFrame(scaler.transform(X),columns=X.columns,index=X.index),Y
# x1,y1=MinMaxScalarNorm(train.drop(columns=['Strength']),train['Strength'])

In [157]:
# train = pd.concat([x1, train['Strength']], axis=1)

In [158]:
from sklearn.model_selection import GridSearchCV
from sklearn.metrics import r2_score

In [159]:
from sklearn.model_selection import train_test_split

X_train, X_val, y_train, y_val = train_test_split(
    train.drop(columns=['Strength']),
    train['Strength'],
    test_size=0.2,
    random_state=42
)

In [160]:
train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 824 entries, 0 to 823
Data columns (total 11 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Cement                  824 non-null    float64
 1   Blast Furnace Slag      824 non-null    float64
 2   Fly Ash                 824 non-null    float64
 3   Water                   824 non-null    float64
 4   Superplasticizer        824 non-null    float64
 5   Coarse Aggregate        824 non-null    float64
 6   Fine Aggregate          824 non-null    float64
 7   Age                     824 non-null    int64  
 8   cement per water        824 non-null    float64
 9   Cement Impurity Factor  824 non-null    float64
 10  Strength                824 non-null    float64
dtypes: float64(10), int64(1)
memory usage: 70.9 KB


In [161]:
train

,Cement,Blast Furnace Slag,Fly Ash,Water,Superplasticizer,Coarse Aggregate,Fine Aggregate,Age,cement per water,Cement Impurity Factor,Strength
0,310.0,0.0,0.0,192.0,0.0,970.0,850.0,7,1.614583,-1.494973,14.99
1,286.3,200.9,0.0,144.7,11.2,1004.6,803.7,56,1.978576,14.926196,72.99
2,362.6,189.0,0.0,164.9,11.6,944.7,755.8,3,2.198908,7.749160,35.30
3,439.0,177.0,0.0,186.0,11.1,884.9,707.9,91,2.360215,5.005339,73.30
4,136.0,162.0,126.0,172.0,10.0,923.0,764.0,28,0.790698,8.589054,29.07
...,...,...,...,...,...,...,...,...,...,...,...
819,116.0,173.0,0.0,192.0,0.0,909.8,891.9,28,0.604167,-2.476372,22.35
820,288.0,192.0,0.0,192.0,0.0,932.0,717.8,28,1.500000,7.485427,38.80
821,237.5,237.5,0.0,228.0,0.0,932.0,594.0,7,1.041667,2.748946,26.26
822,190.0,190.0,0.0,228.0,0.0,932.0,670.0,90,0.833333,-15.306242,42.33


In [162]:
from xgboost import XGBRegressor

xgb = XGBRegressor(
    n_estimators=30, learning_rate=0.1,
    max_depth=5, subsample=0.8, colsample_bytree=0.8,
    reg_alpha=0.1, reg_lambda=1.0, random_state=42
)


model=xgb.fit(X_train, y_train)
y_pred=model.predict(X_val)
score=round(r2_score(y_pred=y_pred,y_true=y_val),3)*100
score

90.3

In [163]:
# from sklearn.model_selection import train_test_split, cross_val_score
# from sklearn.preprocessing import StandardScaler
# from sklearn.pipeline import make_pipeline
# from sklearn.linear_model import Ridge, Lasso, ElasticNet
# from sklearn.svm import SVR
# from sklearn.ensemble import RandomForestRegressor, GradientBoostingRegressor
# import numpy as np

# X = train.drop(columns=['Strength'])
# y = train['Strength']

# X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# models = {
#     "Ridge": Ridge(alpha=1.0),
#     "Lasso": Lasso(alpha=0.01),
#     "ElasticNet": ElasticNet(alpha=0.1, l1_ratio=0.5),
#     "SVR": make_pipeline(StandardScaler(), SVR(C=1.0, kernel='rbf', gamma='scale')),
#     "RandomForest": RandomForestRegressor(
#         n_estimators=200,
#         max_depth=8,
#         min_samples_leaf=5,
#         random_state=42
#     ),
#     "GradientBoosting": GradientBoostingRegressor(
#         n_estimators=300,
#         learning_rate=0.05,
#         max_depth=3,
#         subsample=0.8,
#         random_state=42
#     )
# }
# mod=""
# for name, model in models.items():
#     scores = cross_val_score(model, X_train, y_train, cv=5, scoring='r2')
#     print(f"{name:15s}, {np.mean(scores):.3f}")


In [164]:
# test=test.fillna(test.mean())

<h2 dir=rtl align=right style="line-height:200%;font-family:vazirmatn;color:#0099cc">
<font face="vazirmatn" color="#0099cc">
آموزش مدل
</font>
</h2>

<p dir=rtl style="direction: rtl;text-align: justify;line-height:200%;font-family:vazirmatn;font-size:medium">
<font face="vazirmatn" size=3>
    حال که داده را پاکسازی کرده و احتمالا ویژگی‌هایی را به آن افزوده یا از آن حذف کرده‌اید، وقت آن است که مدلی آموزش دهید که بتواند متغیر هدف این مسئله را پیش‌بینی کند. در این قسمت استفاده از هر نوع مدل مربوط به یادگیری ماشین کلاسیک یا یادگیری عمیق مجاز می‌باشد.
</font>
</p>

<p dir=rtl style="direction: rtl;text-align: justify;line-height:200%;font-family:vazirmatn;font-size:medium">
<font face="vazirmatn" size=3>
در این بخش، وظیفه شما توسعه مدلی است که با استفاده از داده‌های آموزش و بدون دسترسی به برچسب‌ها (labels) بتواند تشخیص دهد که هر نمونه از داده‌های آزمون، یک نمونه عادی است یا یک نمونه ناهنجار.
این مدل باید تنها با تکیه بر الگوها و توزیع ویژگی‌ها در داده‌های آموزش، رفتار نرمال سیستم را یاد بگیرد و بر اساس آن، میزان ناهنجاری هر نمونه از داده‌های آزمون را تخمین بزند.
</font>
</p>

In [165]:
# model=models['RandomForest']
# model.fit(X_train,y_train)

In [166]:
test.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 206 entries, 0 to 205
Data columns (total 10 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Cement                  206 non-null    float64
 1   Blast Furnace Slag      206 non-null    float64
 2   Fly Ash                 206 non-null    float64
 3   Water                   206 non-null    float64
 4   Superplasticizer        206 non-null    float64
 5   Coarse Aggregate        206 non-null    float64
 6   Fine Aggregate          206 non-null    float64
 7   Age                     206 non-null    int64  
 8   cement per water        206 non-null    float64
 9   Cement Impurity Factor  206 non-null    float64
dtypes: float64(9), int64(1)
memory usage: 16.2 KB


<h2 dir=rtl align=right style="line-height:200%;font-family:vazirmatn;color:#0099cc">
<font face="vazirmatn" color="#0099cc">
معیار ارزیابی
</font>
</h2>

<p dir=rtl style="direction: rtl; text-align: justify; line-height:200%; font-family:vazirmatn; font-size:medium">
<font face="vazirmatn" size=3>
    معیاری که برای ارزیابی عملکرد مدل انتخاب کرده‌ایم، <code>r2_score</code> نام دارد.
    <br>
    این معیار، سنجه ارزیابی کیفیت مدل شماست. به عبارت بهتر در سامانه داوری هم از همین معیار برای نمره‌دهی استفاده شده است.
    <br>
    پیشنهاد می‌شود با توجه به این معیار، عملکرد مدل خود را بر روی مجموعه‌ی آموزش یا اعتبارسنجی ارزیابی کنید.
</font>
</p>

<p dir=rtl style="direction: rtl; text-align: justify; line-height:200%; font-family:vazirmatn; font-size:medium">
<font color="red"><b color='red'>توجه:</b></font>
<font face="vazirmatn" size=3>
    برای دریافت نمره از این سوال لازم است تا دقت مدل شما از آستانه‌ی ۰.۶ بیشتر باشد.
    در صورتی که دقت مدل شما از ۰.۶ کمتر باشد نمره شما 
    <b>صفر</b>خواهد بود.
</font>
</p>

<h2 dir=rtl align=right style="line-height:200%;font-family:vazir;color:#0099cc">
<font face="vazir" color="#0099cc">
 پیش‌بینی برای داده تست و خروجی
</font>
</h2>

<p dir=rtl style="direction: rtl;text-align: right;line-height:200%;font-family:vazir;font-size:medium">
<font face="vazir" size=3>
    پیش‌بینی مدل خود بر روی داده‌های آزمایش (<code>test.csv</code>) را در قالب یک <code>dataframe</code> در متغیری با نام <code>submission</code>ذخیره کنید.  این <code>dataframe</code> باید دارای یک ستون با نام <code>Strength</code> باشد که ردیف <code>i</code>ام آن، پیش‌بینی شما برای سطر <code>i</code>ام مجموعه‌داده‌ی آزمون باشد.
</font>
</p>

<center>
<div dir=rtl style="direction: rtl;line-height:200%;font-family:vazir;font-size:medium">
<font face="vazir" size=3>
    
|ستون|توضیحات|
|------|---|
|Strength| مقدار مقاومت بتن هر نمونه آزمایش|
    
</font>

</div>
</center>


<p dir=rtl style="direction: rtl;text-align: right;line-height:200%;font-family:vazir;font-size:medium">
<font face="vazir" size=3>
در این ستون، برای هر نمونه از داده‌های آزمون مشخص می‌کنید که مقدار مقاومت بتن برای آن نمونه آزمون چقدر است.
جدول زیر نمونه‌ای از پنج سطر فایل خروجی مورد انتظار را نشان می‌دهد:
</font>
</p>

<center>
<div dir=rtl style="direction: rtl;line-height:200%;font-family:vazir;font-size:medium">
<font face="vazir" size=3>

|Strength|
|------|
|14.99|
|72.99|
|35.30|
|11.85|
|10.76|

</font>
</div>
</center>

In [167]:
# TODO:
submission = pd.DataFrame({
    "Strength": model.predict(test)
})
submission["Strength"] = submission["Strength"].map(lambda x: f"{x:.2f}")

In [168]:
min(submission.Strength),max(submission.Strength),min(train.Strength),max(train.Strength)


('10.25', '70.75', 2.33, 82.6)

In [169]:
submission

,Strength
0,47.79
1,45.59
2,70.75
3,32.35
4,14.72
...,...
201,56.25
202,27.58
203,35.23
204,63.65


<h2 dir=rtl align=right style="line-height:200%;font-family:vazirmatn;color:#0099cc">
<font face="vazirmatn" color="#0099cc">
<b>سلول جواب‌ساز</b>
</font>
</h2>

<p dir=rtl style="direction: rtl; text-align: justify; line-height:200%; font-family:vazirmatn; font-size:medium">
<font face="vazirmatn" size=3>
    برای ساخته‌شدن فایل <code>result.zip</code> سلول زیر را اجرا کنید. توجه داشته باشید که پیش از اجرای سلول زیر تغییرات اعمال شده در نت‌بوک را ذخیره کرده باشید (<code>ctrl+s</code>) در غیر این صورت، در پایان مسابقه نمره شما به صفر تغییر خواهد کرد.
    <br>
    همچنین اگر از کولب برای اجرای این فایل نوت‌بوک استفاده می‌کنید، قبل از ارسال فایل <code>result.zip</code>، آخرین نسخه‌ی نوت‌بوک خود را دانلود کرده و داخل فایل ارسالی قرار دهید.
</font>

In [172]:
import zipfile
import joblib
import os

if not os.path.exists(os.path.join(os.getcwd(), 'notebook.ipynb')):
    %notebook -e notebook.ipynb

def compress(file_names):
    print("File Paths:")
    print(file_names)
    compression = zipfile.ZIP_DEFLATED
    with zipfile.ZipFile("result.zip", mode="w") as zf:
        for file_name in file_names:
            zf.write('./' + file_name, file_name, compress_type=compression)

submission.to_csv('submission.csv', index=False)
file_names = ['notebook.ipynb', 'submission.csv']
compress(file_names)

File Paths:
['notebook.ipynb', 'submission.csv']
